handlers/media.go

```diff
@@ -0,0 +1,170 @@
+package handlers
+
+import (
+	"context"
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/nyaruka/courier"
+	"github.com/pkg/errors"
+)
+
+type MediaType string
+
+const (
+	MediaTypeImage       MediaType = "image"
+	MediaTypeAudio       MediaType = "audio"
+	MediaTypeVideo       MediaType = "video"
+	MediaTypeApplication MediaType = "application"
+)
+
+type MediaTypeSupport struct {
+	Types    []string
+	MaxBytes int
+}
+
+// Attachment is a resolved attachment
+type Attachment struct {
+	Type        MediaType
+	Name        string
+	ContentType string
+	URL         string
+	Media       courier.Media
+	Thumbnail   courier.Media
+}
+
+// ResolveAttachments resolves the given attachment strings (content-type:url) into attachment objects
+func ResolveAttachments(ctx context.Context, b courier.Backend, attachments []string, support map[MediaType]MediaTypeSupport, allowURLOnly bool) ([]*Attachment, error) {
+	resolved := make([]*Attachment, 0, len(attachments))
+
+	for _, as := range attachments {
+		att, err := resolveAttachment(ctx, b, as, support, allowURLOnly)
+		if err != nil {
+			return nil, err
+		}
+		if att != nil {
+			resolved = append(resolved, att)
+		}
+	}
+
+	return resolved, nil
+}
+
+func resolveAttachment(ctx context.Context, b courier.Backend, attachment string, support map[MediaType]MediaTypeSupport, allowURLOnly bool) (*Attachment, error) {
+	// split into content-type and URL
+	parts := strings.SplitN(attachment, ":", 2)
+	if len(parts) <= 1 || strings.HasPrefix(parts[1], "//") {
+		return nil, errors.Errorf("invalid attachment format: %s", attachment)
+	}
+	contentType, mediaUrl := parts[0], parts[1]
+
+	media, err := b.ResolveMedia(ctx, mediaUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	if media == nil {
+		// if the channel type allows it, we can still use the media URL without being able to resolve it
+		if allowURLOnly {
+			// potentially fix the URL which might not be properly encoded
+			parsedURL, err := url.Parse(mediaUrl)
+			if err == nil {
+				mediaUrl = parsedURL.String()
+			}
+
+			mediaType, _ := parseContentType(contentType)
+			name := filenameFromURL(mediaUrl)
+			return &Attachment{Type: mediaType, Name: name, ContentType: contentType, URL: mediaUrl}, nil
+		} else {
+			return nil, nil
+		}
+	}
+
+	mediaType, _ := parseContentType(media.ContentType())
+	mediaSupport := support[mediaType]
+
+	// our candidates are the uploaded media and any alternates of the same media type
+	candidates := append([]courier.Media{media}, filterMediaByType(media.Alternates(), mediaType)...)
+
+	// narrow down the candidates to the ones we support
+	if len(mediaSupport.Types) > 0 {
+		candidates = filterMediaByContentTypes(candidates, mediaSupport.Types)
+	}
+
+	// narrow down the candidates to the ones that don't exceed our max bytes
+	if mediaSupport.MaxBytes > 0 {
+		candidates = filterMediaBySize(candidates, mediaSupport.MaxBytes)
+	}
+
+	// if we have no candidates, we can't use this media
```

Comment:
I think we need to be backward compatible for the attachments uploaded before we support media as DB objects too

Reply:
I think that has to depend on the channel type - for some channel types it's fine if we have a URL we can't match to a media DB object. For LINE it won't work but I'm fine telling our LINE users to re-upload attachments. I don't think we can automatically backfill every attachment ever uploaded.

Reply:
In the LINE channel I think we resolve attachments with `allowURLOnly = false` and we send whatever attachments we get back which will have an associated `Media` object.